File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/IDCardToolkitObjCApp/IDCardToolkitObjCApp/ViewController/SetNFCdetailsViewController.m
//
//  SetNFCdetailsViewController.m
//  IDCardToolkitObjCApp
//
//  Created by Federal Authority For Identity and Citizenship on 17/10/19.
//  Copyright © 2019 Federal Authority For Identity and Citizenship. All rights reserved.
//

#import "SetNFCdetailsViewController.h"

API_AVAILABLE(ios(13.0))
@interface SetNFCdetailsViewController ()<NFCTagReaderSessionDelegate> {
    CardPublicData *cardPublicData;
    NSString *cardNumberString;
    NSString *dobstring;
    NSString *expireDateString ;
}
@property(nonatomic)NFCTagReaderSession *session;
@property(nonatomic, assign)BOOL isCardV3;
@end

@implementation SetNFCdetailsViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setIsCardV3:NO];
    self.title=@"NFC Card Details View";
      self.navigationItem.hidesBackButton = YES;
      self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[[UIImage imageNamed:@"backbutton"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal]
                                                                               style:UIBarButtonItemStylePlain
                                                                              target:self
                                                                              action:@selector(goBack)];
      
    self.nfcReadpublicdataButton.layer.cornerRadius=22.0;
    [self.nfcReadpublicdataButton.layer setShadowOffset:CGSizeMake(5, 5)];
    [self.nfcReadpublicdataButton.layer setShadowColor:[[UIColor lightGrayColor] CGColor]];
    [self.nfcReadpublicdataButton.layer setShadowOpacity:0.5];
    //self.nfcReadpublicdataButton.layer.borderWidth=1.0;
    
    [self setNumberPad:self.cardNumberText];
    [self setNumberPad:self.dobYYtext];
    [self setNumberPad:self.dobMMText];
    [self setNumberPad:self.dobDDText];
    [self setNumberPad:self.expiryDateYYText];
    [self setNumberPad:self.expiryDateMMText];
    [self setNumberPad:self.expiryDateDDText];
  
      AppDelegate *aDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
      self.model = aDelegate.model;
    
    // Do any additional setup after loading the view.
    [self showCardVersionsAlert];
}

-(void)showCardVersionsAlert {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Card Version" message:@"Please select your card version." preferredStyle:UIAlertControllerStyleAlert];
   
    //Button 1
    UIAlertAction *vThreeCard = [UIAlertAction actionWithTitle:@"V3/V4 Card" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
    //code for performing acctions to be excuted when click Button1
        if (@available(iOS 13.0, *)) {
            [self setIsCardV3:YES];
              self.session = [[NFCTagReaderSession alloc]
                              initWithPollingOption:NFCPollingISO14443 delegate:self queue:nil];
               self.session.alertMessage =  @"Hold your iPhone near an NFC.";
                  [ self.session beginSession];
            
          } else {
               [AlertView showAlertTitle:ALERT withMessage:NFC_NOT_SUPPORT onView:self];
          }
    }];
    
    [alert addAction:vThreeCard];
    
    UIAlertAction *vTwoCard = [UIAlertAction actionWithTitle:@"V2 Card" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
    //code for performing acctions to be excuted when click Button1
        [self setIsCardV3:NO];
    }];
    [alert addAction:vTwoCard];
    [self presentViewController:alert animated:YES completion:^{
    //code for performing acctions to be excuted when UIAlertController is started
   
    }];
}

- (IBAction)nfcReadPublicDataButtonAction:(id)sender {
    [self setIsCardV3:NO];
    cardNumberString = self.cardNumberText.text;
    dobstring = [NSString stringWithFormat:@"%@%@%@",self.dobYYtext.text,self.dobMMText.text,self.dobDDText.text];
    expireDateString = [NSString stringWithFormat:@"%@%@%@",self.expiryDateYYText.text,self.expiryDateMMText.text,self.expiryDateDDText.text];
    
      if (@available(iOS 13.0, *)) {
          
            self.session = [[NFCTagReaderSession alloc]
                            initWithPollingOption:NFCPollingISO14443 delegate:self queue:nil];
             self.session.alertMessage =  @"Hold your iPhone near an NFC.";
                [ self.session beginSession];
          
        } else {
             [AlertView showAlertTitle:ALERT withMessage:NFC_NOT_SUPPORT onView:self];
        }
}
#pragma mark NFC Tag Readger Methods

- (void)tagReaderSessionDidBecomeActive:(NFCTagReaderSession *)session  API_AVAILABLE(ios(13.0)){
    NSLog(@"tagReaderSessionDidBecomeActive");
    Toolkit *toolkit = self.model.toolkitShared;
               
    if (toolkit==nil) {
        [session invalidateSession];
    }
}

- (void)tagReaderSession:(NFCTagReaderSession *)session didInvalidateWithError:(NSError *)error  API_AVAILABLE(ios(13.0)){
    NSLog(@"readerSession:didInvalidateWithError: (%@)", [error localizedDescription]);
    [AlertView showAlertTitle:ALERT withMessage:[error localizedDescription] onView:self];

}

- (void)tagReaderSession:(NFCTagReaderSession *)session didDetectTags:(NSArray<__kindof id<NFCTag>> *)tags   API_AVAILABLE(ios(13.0)){
     NSLog(@"readerSession:didDetectTags");
    if ([tags count] > 1) {
        [session setAlertMessage:@"More than 1 tag is detected, please try again"];
        [session restartPolling];
        return;
    }
    
     id<NFCTag> firstTag =  tags[0];
     NSLog(@"firstTag %@",firstTag);
    
    if (firstTag.type == NFCTagTypeMiFare) {
        [session setAlertMessage:@"A tag that is not iso7816 is detected, please try again with tag iso7816"];
        NSLog(@"session restartPolling");
         [session restartPolling];
    }
    
    NSString *requestId=[Utils generateSecureKey];

    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^(void){
    
    if (@available(iOS 13.0, *)) {
        if (firstTag.type == NFCTagTypeISO7816Compatible) {
            id<NFCISO7816Tag> iso7816Tag = [firstTag asNFCISO7816Tag];
            @try {
        
             Toolkit *toolkit = self.model.toolkitShared;
            
            if (toolkit==nil) {
                [session invalidateSession];
            }
        
            [toolkit setNfcTag:session tag:iso7816Tag];
            
            CardReader *cardreader = [toolkit getReaderWithEmiratesId];

            if (cardreader==nil) {
                [session invalidateSession];
            }
            if (!cardreader.isConnected) {
                 [cardreader connect];
            }
                
           // yymmdd format

//            [cardreader setNfcAuthenticationParameters:@"000020029" dateOfBirth:@"700112" expiryDate:@"290107"];
        if(![self isCardV3]) {
            [cardreader setNfcAuthenticationParameters:self->cardNumberString dateOfBirth:self->dobstring expiryDate:self->expireDateString];//parameters are not required for V3 card
        }
                 
            dispatch_async(dispatch_get_main_queue(), ^{
                [session setAlertMessage:@"Reading Public Details from card"];
            });// main queue
                  
            self->cardPublicData = [cardreader readPublicData:requestId readnonModifiableData:TRUE readModifiableData:TRUE readPhotography:TRUE readSignatueImage:TRUE readAddress:TRUE];
//
                dispatch_async(dispatch_get_main_queue(), ^{
        
                    if ([self->cardPublicData getIdNumber]  != nil) {
                        NSLog(@"idnumber %@",[self->cardPublicData getIdNumber]);
                        NSLog(@"cardNumber %@",[self->cardPublicData getCardNumber]);
                        [self performSegueWithIdentifier:@"NFCpublicDataView" sender:self];
                    }
                        
                     [session setAlertMessage:@"Reading Completed, session going to close"];
                    [session invalidateSession];
                   });// update UI main queue
              }
            @catch (NSException *exception) {
                 [session invalidateSession];
                  NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
                  [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
              }// catch
              @finally {
              }// finally
        }
     } else {
         dispatch_async(dispatch_get_main_queue(), ^{
            // Fallback on earlier versions
            [AlertView showAlertTitle:ALERT withMessage:NFC_NOT_SUPPORT onView:self];
         });// update UI main queue
      }
   });// background queue
}
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    
    if ([[segue identifier] isEqualToString:@"NFCpublicDataView"])
       {
           self.cardNumberText.text=@"";
           self.dobYYtext.text=@"";
           self.dobMMText.text=@"";
           self.dobDDText.text=@"";
           self.expiryDateYYText.text=@"";
           self.expiryDateMMText.text=@"";
           self.expiryDateDDText.text=@"";
           
           // Get reference to the destination view controller
           NFCPublicDataViewController *detailView = [segue destinationViewController];

           [detailView getPublicaDataClass:cardPublicData];
       }
}
-(void)setNumberPad:(UITextField *)customtextfiled {
 
   customtextfiled.keyboardType=UIKeyboardTypeNumberPad;
    customtextfiled.keyboardAppearance = UIKeyboardAppearanceLight;
    
    UIToolbar* NumberToolbar = [[UIToolbar alloc]initWithFrame:CGRectMake(0, 0, 320, 50)];
       NumberToolbar.barStyle = UIBarStyleBlackTranslucent;
       NumberToolbar.items = @[[[UIBarButtonItem alloc]initWithTitle:@"Cancel" style:UIBarButtonItemStyleDone target:self action:@selector(doneWithNumberPad)],
                               [[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil],
                               [[UIBarButtonItem alloc]initWithTitle:@"Done" style:UIBarButtonItemStyleDone target:self action:@selector(doneWithNumberPad)]];
       [NumberToolbar sizeToFit];
       customtextfiled.inputAccessoryView = NumberToolbar;
}
-(void)doneWithNumberPad {
    
    [self.cardNumberText resignFirstResponder];
    [self.dobYYtext resignFirstResponder];
    [self.dobMMText resignFirstResponder];
    [self.dobDDText resignFirstResponder];
    [self.expiryDateYYText resignFirstResponder];
    [self.expiryDateMMText resignFirstResponder];
    [self.expiryDateDDText resignFirstResponder];
}
- (void)goBack {
    [self.navigationController popViewControllerAnimated:YES];
}
/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/


@end

File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/ICASwiftApp/ICASwiftApp/SetNFCdetailsViewController.swift
//
//  SetNFCdetailsViewController.swift
//  ICASwiftApp
//
//  Created by Digital Trust Tech on 06/05/22.
//

import UIKit
import CoreNFC

class SetNFCdetailsViewController: UIViewController, NFCTagReaderSessionDelegate,UITextFieldDelegate {
    
    @IBOutlet weak var cardNumberText: UITextField!
    @IBOutlet weak var dobYYtext: UITextField!
    @IBOutlet weak var dobMMText: UITextField!
    @IBOutlet weak var dobDDText: UITextField!
    @IBOutlet weak var expiryDateYYText: UITextField!
    @IBOutlet weak var expiryDateMMText: UITextField!
    @IBOutlet weak var expiryDateDDText: UITextField!
    @IBOutlet weak var nfcReadpublicdataButton: UIButton!
    
    @IBOutlet weak var view_Roundedcorners: UIView!
    var cardNumberString:String?
    var dobstring:String?
    var expireDateString:String?
    var session: NFCTagReaderSession?
    
    var documentNumber = ""
    var dob_YY = ""
    var dob_MM = ""
    var dob_DD = ""
    var expiryDate_YY = ""
    var expiryDate_MM = ""
    var expiryDate_DD = ""
    
    var isMrzSelected = false

    override func viewDidLoad() {
        super.viewDidLoad()
        
        
        self.view_Roundedcorners.cornered_View([.layerMaxXMaxYCorner, .layerMinXMaxYCorner], radius: 20, borderColor: .red, borderWidth: 0)
        
        // Do any additional setup after loading the view.
        
        nfcReadpublicdataButton.layer.cornerRadius = 22.0
        nfcReadpublicdataButton.layer.shadowOffset = CGSize(width: 5.0, height: 5.0)
        nfcReadpublicdataButton.layer.shadowColor = UIColor.lightGray.cgColor
        nfcReadpublicdataButton.layer.shadowOpacity = 0.5
        
        self.setNumberPad(cardNumberText);
        self.setNumberPad(dobYYtext);
        self.setNumberPad(dobMMText);
        self.setNumberPad(dobDDText);
        self.setNumberPad(expiryDateYYText);
        self.setNumberPad(expiryDateMMText);
        self.setNumberPad(expiryDateDDText);
        
        dobYYtext.delegate = self
        dobMMText.delegate = self
        dobDDText.delegate = self
        expiryDateYYText.delegate = self
        expiryDateMMText.delegate = self
        expiryDateDDText.delegate = self
        
        dobYYtext.addTarget(self, action: #selector(self.textFieldDidChange(textField:)), for: UIControl.Event.editingChanged)
        dobMMText.addTarget(self, action: #selector(self.textFieldDidChange(textField:)), for: UIControl.Event.editingChanged)
        dobDDText.addTarget(self, action: #selector(self.textFieldDidChange(textField:)), for: UIControl.Event.editingChanged)
        expiryDateYYText.addTarget(self, action: #selector(self.textFieldDidChange(textField:)), for: UIControl.Event.editingChanged)
        expiryDateMMText.addTarget(self, action: #selector(self.textFieldDidChange(textField:)), for: UIControl.Event.editingChanged)
        expiryDateDDText.addTarget(self, action: #selector(self.textFieldDidChange(textField:)), for: UIControl.Event.editingChanged)
        
        
        if isMrzSelected == true{
            
            cardNumberText.text = documentNumber
            dobYYtext.text = dob_YY
            dobMMText.text = dob_MM
            dobDDText.text = dob_DD
            
            expiryDateYYText.text = expiryDate_YY
            expiryDateMMText.text = expiryDate_MM
            expiryDateDDText.text = expiryDate_DD
            
            DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) {
                self.cardNumberString = self.cardNumberText.text
                
                self.dobstring = "\(self.dobYYtext.text!)\(self.dobMMText.text!)\(self.dobDDText.text!)"
                self.expireDateString = "\(self.expiryDateYYText.text!)\(self.expiryDateMMText.text!)\(self.expiryDateDDText.text!)"
                
                if #available(iOS 13.0, *) {
                    self.session = NFCTagReaderSession(pollingOption: .iso14443, delegate: self, queue: nil)
                    self.session?.alertMessage = "Hold your iPhone near an NFC."
                    self.session?.begin()
                } else {
                    // Fallback on earlier versions
                    AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.NFC_Not_Support,onView: self)
                }
            }
        }
    }
    @objc func textFieldDidChange(textField: UITextField){
        let text = textField.text
        if  text?.count == 2 {
            switch textField{
            case dobYYtext:
                dobMMText.becomeFirstResponder()
                
            case dobMMText:
                dobDDText.becomeFirstResponder()
                
            case dobDDText:
                expiryDateYYText.becomeFirstResponder()
                
            case expiryDateYYText:
                expiryDateMMText.becomeFirstResponder()
                
            case expiryDateMMText:
                expiryDateDDText.becomeFirstResponder()
                
            case expiryDateDDText:
                expiryDateDDText.resignFirstResponder()
            default:
                break
            }
        }
        if  text?.count == 0 {
            switch textField{
                
            case dobYYtext:
                dobYYtext.becomeFirstResponder()
                
            case dobMMText:
                dobYYtext.becomeFirstResponder()
                
            case dobDDText:
                dobMMText.becomeFirstResponder()
                
            case expiryDateYYText:
                dobDDText.becomeFirstResponder()
                
            case expiryDateMMText:
                expiryDateYYText.becomeFirstResponder()
                
            case expiryDateDDText:
                expiryDateMMText.becomeFirstResponder()
                
            default:
                break
            }
        }
        else{
        }
    }
    
    
    @IBAction func btn_BackAction(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }
    
    @IBAction func nfcReadPublicDataButtonAction(_ sender: Any) {
        
        cardNumberString = cardNumberText.text
        //        dobstring = "\(dobYYtext.text)\(dobMMText.text)\(dobDDText.text)"
        //        expireDateString = "\(expiryDateYYText.text)\(expiryDateMMText.text)\(expiryDateDDText.text)"
        dobstring = "\(dobYYtext.text!)\(dobMMText.text!)\(dobDDText.text!)"
        expireDateString = "\(expiryDateYYText.text!)\(expiryDateMMText.text!)\(expiryDateDDText.text!)"
        
        if toolkit == nil {
            DispatchQueue.main.async {
                self.view.makeToast(Constants.toolkitNotInitilized)
            }
            return
        }//if
        
        if #available(iOS 13.0, *) {
            self.session = NFCTagReaderSession(pollingOption: .iso14443, delegate: self, queue: nil)
            self.session?.alertMessage = "Hold your iPhone near an NFC."
            self.session?.begin()
        } else {
            // Fallback on earlier versions
            AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.NFC_Not_Support,onView: self)
        }
    }
    
    @available(iOS 13.0, *)
     func tagReaderSessionDidBecomeActive(_ session: NFCTagReaderSession) {
         print("tagReaderSessionDidBecomeActive")
         
         if toolkit == nil {
             DispatchQueue.main.async {
                 self.view.makeToast(Constants.toolkitNotInitilized)
                 session.invalidate()
             }
             return
         }//if
        
     }
    
     @available(iOS 13.0, *)
     func tagReaderSession(_ session: NFCTagReaderSession, didInvalidateWithError error: Error) {
         print("readerSession:didInvalidateWithError: \(error.localizedDescription)")
         AlertView.showAlertTitle(Constants.Alert, withMessage:error.localizedDescription,onView: self)
     }
    @available(iOS 13.0, *)
    func tagReaderSession(_ session: NFCTagReaderSession, didDetect tags: [NFCTag]) {
        print("readerSession:didDetectTags")
        
        if tags.count > 1 {
            let retryInterval = DispatchTimeInterval.milliseconds(500)
            session.alertMessage = "More than 1 tag is detected, please try again"
            DispatchQueue.global().asyncAfter(deadline: .now() + retryInterval, execute: {
                session.restartPolling()
            })
            return
        }
        
  
        let firstTag = tags.first!
        print("firstTag \(firstTag)")
        
        guard case .iso7816(let iso7816Tag) = firstTag else {
            let retryInterval = DispatchTimeInterval.milliseconds(500)
            session.alertMessage = "A tag that is not iso7816 is detected, please try again with tag iso7816."
            DispatchQueue.global().asyncAfter(deadline: .now() + retryInterval, execute: {
                session.restartPolling()
            })
            return
        }
        
        print("iso7816Tag identifier \(iso7816Tag.identifier)" )
        print("iso7816Tag initialSelectedAID \(iso7816Tag.initialSelectedAID)" )
        print("iso7816Tag historicalBytes \(iso7816Tag.historicalBytes)" )
        print("iso7816Tag applicationData \(iso7816Tag.applicationData)" )

        if toolkit == nil {
            DispatchQueue.main.async {
                self.view.makeToast(Constants.toolkitNotInitilized)
                session.invalidate()
            }
            return
        }//if
        
        print("session \(session)")
        print("iso7816Tag \(iso7816Tag)")
        let requestId = Utils.generateSecureKey()
        
      DispatchQueue.global(qos: .background).async {
            
               do {
                try toolkit?.setNfcTag(session, tag: iso7816Tag)
                
                cardreader = try toolkit?.getReaderWithEmiratesId()
                
                if cardreader == nil {
                    session.invalidate()
                    return
                }//if
                
                if !(cardreader?.isConnected())! {
                    try cardreader?.connect()
                }
                
                 // yymmdd format
                
                print("\(self.cardNumberString!)")
                print("\(self.dobstring!)")
                print("\(self.expireDateString!)")
                try cardreader?.setNfcAuthenticationParameters(self.cardNumberString!, dateOfBirth: self.dobstring!, expiryDate: self.expireDateString!)//parameters are not required for V3 card

            DispatchQueue.main.async {
                session.alertMessage = "Reading Public Details from card"
            }// update UI main queue
                
            cardPublicData = try cardreader?.readPublicData(requestId!, readNonModifiableData: true, readModifiableData: true, readPhotography: true, readSignatueImage: true, readAddress: true)
                 
              DispatchQueue.main.async {
                
                if cardPublicData.getIdNumber() != nil {
                    print("getIdNumber \(cardPublicData.getIdNumber())")
                    print("getCardNumber \(cardPublicData.getCardNumber())")
                    self.performSegue(withIdentifier: "NFCpublicDataView", sender: self)
                }
                session.alertMessage = "Reading Completed, session going to close"
                session.invalidate()
             }// update UI main queue
                                
             }//do
        catch let error as NSError{
            DispatchQueue.main.async {
                session.invalidate()
                let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                utils.dismissProgressBar()
            }// update UI main queue
         }//catch
        }//background queue
    }
    
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {

        let segueID = segue.identifier
        if(segueID == "NFCpublicDataView") {
            
            self.cardNumberText.text = ""
            self.dobYYtext.text = ""
            self.dobMMText.text = ""
            self.dobDDText.text = ""
            self.expiryDateYYText.text = ""
            self.expiryDateMMText.text = ""
            self.expiryDateDDText.text = ""
            
            let detailView = segue.destination as! NFCPublicDataViewController
            detailView.getPublicaDataClass(cardPublicData)
        }
    }
    
    
    func setNumberPad(_ customtextfiled:UITextField) {
        
        customtextfiled.keyboardType = .numberPad
        customtextfiled.keyboardAppearance = .light
        let confirmPinNumbertoolbar = UIToolbar(frame: CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(320), height: CGFloat(50)))
        confirmPinNumbertoolbar.barStyle = .blackTranslucent
        confirmPinNumbertoolbar.items = [UIBarButtonItem(title: "Cancel", style: .done, target: self, action: #selector(doneWithNumberPad)), UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil), UIBarButtonItem(title: "Done", style: .done, target: self, action: #selector(doneWithNumberPad))]
        customtextfiled.inputAccessoryView = confirmPinNumbertoolbar
    }
    
    
    @objc func doneWithNumberPad() {
        self.cardNumberText.resignFirstResponder()
        self.dobYYtext.resignFirstResponder()
        self.dobMMText.resignFirstResponder()
        self.dobDDText.resignFirstResponder()
        self.expiryDateYYText.resignFirstResponder()
        self.expiryDateMMText.resignFirstResponder()
        self.expiryDateDDText.resignFirstResponder()
        
    }
    
    @objc func goBack() {
        navigationController?.popViewController(animated: true)
    }
    

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}
